cranelift/zkasm_data/benchmarks/MVP.md

```diff
@@ -0,0 +1,141 @@
+
+# MVP Implementation
+
+## Instrumenting zkASM
+
+### Conditionality
+
+Choose one of the following options to achieve conditionality. For the MVP choice depends mainly on ease of setup.
+
+- A Rust feature to use conditional compilation and optional dependencies.
+  - If instrumentation becomes involved this might be helpful to leave the compiler unaffected when instrumentation is not enabled.
+- Runtime checks to verify whether instrumentation is enabled.
+
+### What to log and why?
+
+#### Hot instructions: `Inst` execution
+
+To build a histogram of executed instructions. This helps to identify instructions for which an optimization of the emitted zkASM may have a high impact as it compounds over many executions.
+
+#### Hot labels: executed labels
+
+To build a histogram of executed labels. This helps to identify:
+
+- Labels most frequently executed which can hint at optimizations of the original program that is compiled to zkASM.
+- By looking at the body of the labels in the generated zkASM, it can help to find instruction sequences for which peephole optimizations might have a high impact as they compound over many executions.
+
+## Aspirational insights to gain while instrumenting zkASM
+
+- Try to get an idea of how to instrument instructions such that they can be assigned to the label they are nested under.
+- Try to understand how call hierarchies are represented in cranelift.
+
+### Where to log?
+
+#### `Inst` execution
+
+In `Inst::emit()` [before matching](https://github.com/near/wasmtime/blob/9dc141e0181e5c1121cf1daa42d80ce26468757b/cranelift/codegen/src/isa/zkasm/inst/emit.rs#L406) the variant.
+
+#### Label execution
+
+- Execution of zkASM labels emitted for `Inst::Label` is handled via logs for `Inst` execution (see above).
+- Logs for labels corresponding to functions are inserted when the label is created in [generate_zkasm\(\)](https://github.com/near/wasmtime/blob/9dc141e0181e5c1121cf1daa42d80ce26468757b/cranelift/filetests/src/test_zkasm.rs#L220).
+
+### What to log?
+
+To build the histograms, execution counters according to the following schema shall be collected:
+
+```json
+{
+    "labels": {
+        "functions": {
+            "function_1": 1,
+            "function_2": 14
+        },
+        "others": {
+            "label_1_1": 4,
+            "label_1_2": 8,
+            "label_2_1": 4
+        }
+    },
+    "instructions": {
+        "AluRRR": 12,
+        "Load": 8,
+        "Store": 8
+    }
+}
+```
+
+### How to log
+
+#### Variant 1: `zkevm-proverjs` helpers can maintain state
+
+If `zkevm-proverjs` helpers can maintain state during the execution of a program, above data structure can live in state. Then instrumented zkASM would call helpers like:
+
+```js
+$${count("labels", "functions", "function_2")}
+
+$${count("instructions", null, "Load")}
+```
+
+Before program termination, the data structure is logged.
+
+#### Variant 2: `zkevm-proverjs` helpers cannot maintain state
```

Comment:
We should be able to maintain state during the execution. E.g. see https://github.com/0xPolygonHermez/zkevm-proverjs/blob/07d92991778be82f66d8be55a3862d2843007221/src/sm/sm_main/sm_main_exec.js#L2751